src/Infrastructure.VS/DEBUG_ONLY_VsCookbookThreadingChecker.cs

```diff
@@ -0,0 +1,154 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+
+namespace SonarLint.VisualStudio.Infrastructure.VS
+{
+    // ******************************************************
+    // **
+    // **  Not be to used in shipping production code !
+    // **
+    // **  Only use for testing whether code is free-threaded
+    // **  in local dev environments.
+    // **
+    // ******************************************************
+
+
+    /* From the VS threading cookbook: https://github.com/microsoft/vs-threading/blob/main/doc/cookbook_vs.md
+ 
+        "Important places to be thread-safe include:
+
+        1. Code that runs in the MEF activation code paths (e.g. importing constructors, OnImportsSatisfied callbacks,
+           and any code called therefrom).
+        2. Code that may run on a background thread, especially when run within an async context.
+        3. Static field initializers and static constructors."
+
+
+        Instructions for use:
+        ---------------------
+        * Decide which code you want to test
+        * Call the code to be tested using one of the helper methods in this class
+        * Execute the code in VS
+        
+        If VS deadlocks the code is not free-threaded.
+
+        From our point of view, cases (1) and (2) above are most relevant.
+
+        Any code that needs the UI thread is by definition not free-threaded (e.g. to request a VS service)
+        is by definition NOT free-threaded, so it will definitely deadlock when using these helper methods.
+
+    */
+    public static class DEBUG_ONLY_VsCookbookThreadingChecker
+    {
+        /// <summary>
+        /// If this method is called on the UI thread it will check
+        /// whether the operation _requires_ the UI thread.
+        /// If the method is called from a background thread it will just execute the operation.
+        /// </summary>
+        public static void IfOnUIThreadCheckIfRequiresMainThread(Action operationToTest)
```

Comment:
There are four methods:

* _IfOnUIThreadCheckIfRequiresMainThread(Action)_  : only does the check if the caller is on the UI thread.
* _CheckIfRequiresMainThread(Action)_  : does the check

... and equivalent methods for `Func<T>`.